src/psyclone/psyir/transformations/intrinsics/maxval2code_trans.py

```diff
@@ -0,0 +1,204 @@
+# BSD 3-Clause License
+#
+# Copyright (c) 2023, Science and Technology Facilities Council
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: R. W. Ford, STFC Daresbury Lab
+
+'''Module providing a transformation from a PSyIR MAXVAL intrinsic to
+PSyIR code. This could be useful if the MAXVAL operator is not
+supported by the back-end, the required parallelisation approach, or
+if the performance in the inline code is better than the intrinsic.
+
+'''
+from psyclone.psyir.nodes import (
+    Reference, Assignment, IfBlock, BinaryOperation, ArrayReference,
+    IntrinsicCall)
+from psyclone.psyir.transformations.intrinsics.mms_base_trans import (
+    MMSBaseTrans)
+
+
+class Maxval2CodeTrans(MMSBaseTrans):
+    '''Provides a transformation from a PSyIR MAXVAL Operator node to
+    equivalent code in a PSyIR tree. Validity checks are also
+    performed.
+
+    If MAXVAL contains a single positional argument which is an array,
+    the maximum value of all of the elements in the array is returned
+    in the the scalar R.
+
+    .. code-block:: python
+
+        R = MAXVAL(ARRAY)
+
+    For example, if the array is two dimensional, the equivalent code
+    for real data is:
+
+    .. code-block:: python
+
+        R = TINY(R)
+        DO J=LBOUND(ARRAY,2),UBOUND(ARRAY,2)
+          DO I=LBOUND(ARRAY,1),UBOUND(ARRAY,1)
+            IF R < ARRAY(I,J) THEN
```

Comment:
Hopefully you meant `.. codeblock:: fortran` as that seems to be what we use so I have continued with this. You are right that it has become Fortran code so I've added prentheses where required. I've also updated the minval and sum transformation code.

Reply:
I did and it looks nice now thanks. There must be a syntax error though because the rendered doc contains:
![image](https://github.com/stfc/PSyclone/assets/17409212/aef87f60-e232-4367-a846-d50e77fb4f50)

Reply:
Oops. Fixed now for all three transformations.